pkg/cli/gettoken/gettoken.go

```diff
@@ -0,0 +1,269 @@
+	cmd.Flags().BoolVar(&o.AutoOpenBrowser, "auto-open-browser", o.AutoOpenBrowser, "Specify browser is automatically opened or not.")
+
+	return cmd
+}
+
+func (o *GetTokenOptions) Complete(f kcmdutil.Factory, cmd *cobra.Command, args []string) error {
+	o.CACertFilename = kcmdutil.GetFlagString(cmd, "certificate-authority")
+	o.InsecureTLS = kcmdutil.GetFlagBool(cmd, "insecure-skip-tls-verify")
+
+	provider := oidc.Provider{
+		IssuerURL:    o.IssuerURL,
+		ClientID:     o.ClientID,
+		ClientSecret: o.ClientSecret,
+		ExtraScopes:  o.ExtraScopes,
+		UsePKCE:      true,
+	}
+
+	authenticator, err := oidc.NewAuthenticator(context.TODO(), provider, "", o.CACertFilename, o.InsecureTLS)
+	if err != nil {
+		return fmt.Errorf("oidc authenticator error %v", err)
+	}
+
+	o.authenticator = authenticator
+	o.tokenCache = &tokencache.Repository{}
+	o.credWriter = credwriter.NewWriter(o.IOStreams)
+
+	if kcd := os.Getenv("KUBECACHEDIR"); kcd != "" {
+		o.tokenCacheDir = filepath.Join(kcd, "oc")
+	} else {
+		o.tokenCacheDir = filepath.Join(homedir.HomeDir(), ".kube", "cache", "oc")
+	}
+
+	return nil
+}
+
+func (o *GetTokenOptions) Validate() error {
+	if o.IssuerURL == "" {
+		return fmt.Errorf("--issuer-url is required")
+	}
+	if o.ClientID == "" {
+		return fmt.Errorf("--client-id is required")
+	}
+
+	return nil
+}
+
+// Run starts the authentication flow with a caching and refreshing capability.
+// If refresh token is found, it tries to use it to get a valid id token from
+// external OIDC issuer. If not, it forces user to log in.
+func (o *GetTokenOptions) Run() error {
+	tokenCacheKey := tokencache.Key{
+		IssuerURL:      o.IssuerURL,
+		ClientID:       o.ClientID,
+		ExtraScopes:    o.ExtraScopes,
+		CACertFilename: o.CACertFilename,
+		SkipTLSVerify:  o.InsecureTLS,
+	}
+
+	tokenSet, err := o.tokenCache.FindByKey(o.tokenCacheDir, tokenCacheKey)
+	if err != nil {
+		return err
+	}
+	alreadyValid, idToken, refreshToken, err := o.getToken(context.TODO(), tokenSet)
+	if err != nil {
+		return err
+	}
+
+	idTokenClaims, err := jwt.DecodeWithoutVerify(idToken)
+	if err != nil {
+		return fmt.Errorf("you got an invalid token: %w", err)
+	}
+
+	if !alreadyValid {
+		err = o.tokenCache.Save(o.tokenCacheDir, tokenCacheKey, tokencache.Set{
+			IDToken:      idToken,
+			RefreshToken: refreshToken,
+		})
+		if err != nil {
+			return fmt.Errorf("could not write into the token cache")
+		}
+	}
+
+	if err := o.credWriter.Write(idToken, idTokenClaims.Expiry); err != nil {
+		return fmt.Errorf("could not write the token to client-go: %w", err)
+	}
+	return nil
+}
+
+func (o *GetTokenOptions) getToken(ctx context.Context, cache *tokencache.Set) (bool, string, string, error) {
+	if cache != nil {
+		claims, err := jwt.DecodeWithoutVerify(cache.IDToken)
+		if err != nil {
```

Comment:
This is the justification https://github.com/int128/kubelogin/blob/79a6f8596774d9d1fff2319ddc84a3a87f80fa48/pkg/usecases/authentication/authentication.go#L82-L85, let me know what you think.

Reply:
We don't cache refresh token without id token, so that if there is no id token, that means user manually deleted the id token from cache file. In that case, we can safely return error warning user about removing this caching file altogether.